### ISSUE TITLE
Runaway RAM consumption while streaming to Janus Gateway

### ISSUE THREAD

**Issue by command-tab** [CoSMoSoftware/OBS-studio-webrtc#235]
Greetings! I'm using the current release (m84v23.2 RC 1) of this wonderful OBS Studio fork along with the current Janus Gateway (built from `master`), and this combo is _so close_ to working great for me. However, I'm finding that after streaming for just over 5 minutes, the stream drops, briefly attempts to reconnect, and then OBS crashes/quits without any message. On the Janus side, its logs show:

```[Wed Sep  2 20:50:31 2020] Creating new session: 866527138776996; 0x7fb7300010c0
[Wed Sep  2 20:50:31 2020] Creating new handle in session 866527138776996: 5634162941838489; 0x7fb7300010c0 0x7fb730001730
[Wed Sep  2 20:50:35 2020] [5634162941838489] Creating ICE agent (ICE Full mode, controlled)
[Wed Sep  2 20:50:35 2020] [5634162941838489] The DTLS handshake has been completed
[Wed Sep  2 20:50:35 2020] [janus.plugin.videoroom-0x7fb730001370] WebRTC media is now available
[Wed Sep  2 20:55:41 2020] [WSS-0x7fb72c0008c0] Destroying WebSocket client
[Wed Sep  2 20:55:41 2020] Destroying session 866527138776996; 0x7fb7300010c0
[Wed Sep  2 20:55:41 2020] Detaching handle from JANUS VideoRoom plugin; 0x7fb730001730 0x7fb730001370 0x7fb730001730 0x7fb730001200
[Wed Sep  2 20:55:41 2020] [janus.plugin.videoroom-0x7fb730001370] No WebRTC media anymore; 0x7fb730001730 0x7fb730001200
[Wed Sep  2 20:55:41 2020] [5634162941838489] WebRTC resources freed; 0x7fb730001730 0x7fb7300010c0
[Wed Sep  2 20:55:41 2020] [5634162941838489] Handle and related resources freed; 0x7fb730001730 0x7fb7300010c0
```

You can see at 20:50 the `WebRTC media is now available` when the stream starts, then at 20:55 `Destroying WebSocket client` occurs along with the associated teardown. After several attempts at streaming, it always seem to happen just after the 5 minute mark. Any idea why that might be, and whether the shutdown is server- or client-initiated?

Thanks so much :heart:

**Comment by command-tab** [comment]
Of note: When streaming from the same source machine as before to the same Janus Gateway server as before, but instead using a web-based WebRTC client using `janus.js` library, I don't encounter the 5-minute drop. This leads me to think it's an application issue and not a server side issue.

**Comment by command-tab** [comment]
I just tried another angle to attempt to make the problem smaller: I ran Janus Gateway on the _same machine_ as the one running OBS and updated my prefs to stream to that server from OBS, so there's no traffic even leaving the machine (avoiding any internet issues, etc.), and again OBS drops right after the 5 minute mark, reconnects briefly, then silently quits/crashes. So, the issue doesn't appear to be related to a specific Janus Gateway instance.

I'd love for someone to attempt to reproduce the issue and debug. If it would help, I'm happy to set up a temporary Janus Gateway instance out on the net.

**Comment by sm2x** [comment]
what is your OS and version?
Janus Gateway version?
how did you install it, your own build or via distribution?

**Comment by command-tab** [comment]
I'm streaming from macOS Catalina 10.15.6 (19G73). I tried sending to two separate but very recently built versions of Janus Gateway:

Remote Linux (CentOS 7):

    Janus commit: b26cbb159a6afb59f1ce99f3352eb2ab98199361
    Compiled on:  Mon Aug 31 16:12:25 UTC 2020
    janus 0.10.5

Local macOS (I did a `git pull` on `master` and built this yesterday):

    Janus commit: 1a825c485ec0dc212569b014a834634c1e977962
    Compiled on:  Wed Sep  2 15:38:22 PDT 2020
    janus 0.10.5

**Comment by command-tab** [comment]
Update: This seems to be a memory leak in the Janus (or WebRTC in general?) stream support. When I stream to any Janus Gateway server, OBS memory continuously creeps up at about 25 MB/sec. At 4 minutes into a stream, OBS is using 6.5 GB. Curiously, if I stream to Periscope, this doesn't happen.

**Comment by sm2x** [comment]
yep, have you reported this on Janus already?

**Comment by command-tab** [comment]
Sorry, I didn't write clearly enough. This seems to be a memory leak in _this OBS fork's support for_ Janus or WebRTC in general. It's not the Janus Gateway that's eating up RAM — it's OBS Studio (Millicast):

![image](https://user-images.githubusercontent.com/3069/92986188-daf01080-f46d-11ea-8e0e-4842b8e7f58d.png)

OBS eats up RAM when streaming to a local or remote Janus server, then OBS crashes silently.

**Comment by agouaillard-cosmo** [comment]
Just to be clear, @sm2x is not affiliated with this project in anyway.

@command-tab , if you could host  janus server for us to test against that would be appreciated. I agree that it seems to be a problem with our obs-studio-webrtc. In the mean time, I will check if I see the same behaviour when using it against other webrtc media server or service.  Note, it is expected that you would not have the same behaviour when streaming against ramp services (or Flt) like periscope, as it doe snot use the same webrtc stack.

**Comment by command-tab** [comment]
Thanks! I'll get a Janus server set up to test against.

**Comment by command-tab** [comment]
OK, here's a Janus Gateway server for you to test against:

```
Server Name  wss://janus.command-tab.com:8989
Server Room  1234
Stream Key   adminpwd
```

I've streamed to it and observed the same runaway RAM consumption and eventual crash I saw with the two other Janus servers.

Thanks for looking into this ❤️

**Comment by command-tab** [comment]
Any chance someone could try to reproduce/triage this issue soon-ish? (I ask because I'm paying to host this Janus Gateway server VM on Linode every month, though it's not a ton. The value of the repro and bug fix is worth the cost. Thank you! 🙏 )

**Comment by agouaillard-cosmo** [comment]
Checking.

**Comment by agouaillard-cosmo** [comment]
I can reproduce a problem with RC 2:
- when clicking "start streaming" the button change to connecting and never transition to either "stop streaming" (success) nor "start streaming" (failure).
- I see the runaway memory, with big steps in memory consumption at relatively low frequency.
Likely a memory leak when retrying.

I will rebuild in debug mode and look at what is happening, good catch.

**Comment by agouaillard-cosmo** [comment]
@command-tab You can shut down your instance for now not to incur undue costs. We have to do a few things before we can come back to this. I will let you know.

**Comment by agouaillard-cosmo** [comment]
After discussion with the meet echo team, we agreed on the following:
1. No support for Janus will be provided by default. Bruden of proof (that there is a bug on the OS Side) is on the one(s) opening the tickets.
2. There is no commercial support for this fork of OBS.
3. Janus support will only be done through the WHIP protocol.
4. Commercial support for Janus server side is available [here](https://janus.conf.meetecho.com/support).

**Comment by agouaillard-cosmo** [comment]
NOTE ALEX: address the runaway memory bug anyway, but open a new ticket.
Investigate reconnection flow.
